crates/ruff/src/rules/ruff/rules/mutable_defaults_in_dataclass_fields.rs

```diff
@@ -0,0 +1,207 @@
+///     mutable_default: list[int] = field(default_factory=list)
+/// ```
+///
+/// Alternatively, if you __want__ the shared behaviour, make it more obvious
+/// by assigning it to a module-level variable:
+/// ```python
+/// from dataclasses import dataclass
+///
+/// I_KNOW_THIS_IS_SHARED_STATE = [1, 2, 3, 4]
+///
+/// @dataclass
+/// class A:
+///     mutable_default: list[int] = I_KNOW_THIS_IS_SHARED_STATE
+/// ```
+#[violation]
+pub struct MutableDataclassDefault;
+
+impl Violation for MutableDataclassDefault {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Do not use mutable default values for dataclass attributes")
+    }
+}
+
+/// ## What it does
+/// Checks for function calls in dataclass defaults.
+///
+/// ## Why is it bad?
+/// Function calls are only performed once during definition time. The result
+/// gets reused in all created instances of the dataclass.
+///
+/// ## Examples:
+/// ```python
+/// from dataclasses import dataclass
+///
+/// def creating_list() -> list[]:
+///     return [1, 2, 3, 4]
+///
+/// @dataclass
+/// class A:
+///     mutable_default: list[int] = creating_list()
+///
+/// # also:
+///
+/// @dataclass
+/// class B:
+///     also_mutable_default_but_sneakier: A = A()
+/// ```
+///
+/// Use instead:
+/// ```python
+/// from dataclasses import dataclass, field
+///
+/// def creating_list() -> list[]:
+///     return [1, 2, 3, 4]
+///
+/// @dataclass
+/// class A:
+///     mutable_default: list[int] = field(default_factory=creating_list)
+///
+/// @dataclass
+/// class B:
+///     also_mutable_default_but_sneakier: A = field(default_factory=A)
+/// ```
+///
+/// Alternatively, if you __want__ the shared behaviour, make it more obvious
+/// by assigning it to a module-level variable:
+/// ```python
+/// from dataclasses import dataclass
+///
+/// def creating_list() -> list[]:
+///     return [1, 2, 3, 4]
+///
+/// I_KNOW_THIS_IS_SHARED_STATE = creating_list()
+///
+/// @dataclass
+/// class A:
+///     mutable_default: list[int] = I_KNOW_THIS_IS_SHARED_STATE
+/// ```
+#[violation]
+pub struct FunctionCallInDataclassDefaultArgument {
+    pub name: Option<String>,
+}
+
+impl Violation for FunctionCallInDataclassDefaultArgument {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let FunctionCallInDataclassDefaultArgument { name } = self;
+        if let Some(name) = name {
+            format!("Do not perform function call `{name}` in dataclass defaults")
+        } else {
+            format!("Do not perform function call in dataclass defaults")
+        }
+    }
+}
+
+fn is_mutable_expr(expr: &Expr) -> bool {
+    matches!(
+        &expr.node,
+        ExprKind::List { .. }
+            | ExprKind::Dict { .. }
+            | ExprKind::Set { .. }
+            | ExprKind::ListComp { .. }
+            | ExprKind::DictComp { .. }
+            | ExprKind::SetComp { .. }
+    )
+}
+
+const ALLOWED_FUNCS: &[&[&str]] = &[&["dataclasses", "field"]];
```

Comment:
We may want to make this a setting, to allow users to allow certain functions here. But, we can wait on that, see if it's a real need that comes up in Issues.

Reply:
I actually thought about that (mainly because the bugbear implementation has something similar), but I couldn't come up with something I would really want there. The only thing that is sensible in my mind at least, would be something like "registering" your `NamedTuples` as they are immutable or ignoring the error for one specific function.

So yeah personally I did leave it as "kinda prepared if necessary, but only if needed in the future" :)

Reply:
Well, now that I've revisited that thought, it could also match the `IMMUTABLE_FUNCS` from [here (function_call_argument_default.rs)](https://github.com/charliermarsh/ruff/blob/abaf0a198d9d2cebefd5c029bb271be1a3498f20/crates/ruff/src/rules/flake8_bugbear/rules/function_call_argument_default.rs#L33) + the added dataclass-specific `dataclass.field` one.

I'm not sure if you want to do that as it would probably mean exporting one or the other (or both) and having configuration options for two (very far apart) rule checkers in one of them; or duplicating the array on both sides.

Reply:
Also this would have implications for "user-added" things. As "sharing" the list would mean the user adds a to-ignore function to the flake8-bugbear-config and it gets ignored in the dataclass as well.

Therefore I'm leaning more to a duplicate of the array at the moment.

Reply:
I'm okay to skip for now. Let's see if it comes up. If you _do_ feel like implementing, my gut reaction is just to duplicate for now, but there's not great precedent to follow (it's odd to have two different configuration options for what are really very similar rules, but we also tend to avoid sharing configuration across "sub-linters").